### ISSUE TITLE
std::abort will be called in destructor of runtime

### ISSUE THREAD

**Issue by Trafo** [David-Haim/concurrencpp#118]
We face in our product, that in case we try to shut down the global runtime, doing tasks are running that an exception appeared, which will be catched, and then a std::abort will be called. That is quite unpleasant. Because on the one hand, I cannot debug the exception on the other hand we hard crash here all the time we try to stop all running tasks or try to finish the application.

I will try first on my side to look what happens, but wanted ask before if this bug is known and already in the upcoming newer version fixed?

**Comment by David-Haim** [comment]
Hi there,
I'm not quite sure to which case are you referring? when the runtime is shut down, it shuts down stored executors, and they, in their turn abort suspended coroutines by resuming them and causing them to throw `runtime_shutdown` or `broken_task` exception.
Are you guys catching those exceptions and make the code exit gracefully? 
Can you also add a minimal example that shows the problem?

**Comment by Trafo** [comment]
Hi,
sorry, it was in the end my fault. I shut down the runtime inside of a coroutine, which was impossible because the shutdown method was blocking at the same time the suspension of the coroutine. Because of that an exception was thrown and the destructor of runtime catches this exception and calls std::abort.

Now I fixed this behavior, called runtime::shutdown in another non-blocking thread, and can reproduce an issue I faced quite often. When in debug mode I call shutdown of the runtime during coroutines are working, an assert will be triggered inside of the code which I don't understand.

```
        ~producer_context() noexcept {
            switch (m_status) {
                case result_status::value: {
                    m_storage.object.~type();
                    break;
                }

                case result_status::exception: {
                    m_storage.exception.~exception_ptr();
                    break;
                }

                case result_status::idle: {
                    break;
                }

                default: {
                    assert(false);
                }
            }
        }
```
From my understanding, m_status will be always initialized and can always have these 3 states. But still, a default will be reached in my case. At the moment I try to build a minimal example.

**Comment by Trafo** [comment]
During try to create a minimal example I found not exactly this issue, but a similar one at another place, which is as well weird regarding the usage of an Enum `m_status`:

```
    auto runtime = std::make_shared<concurrencpp::runtime>();
    auto thread_executor = runtime->thread_executor();
    auto result = thread_executor->submit([=] {
        std::cout << "first layer" << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(1));
        thread_executor->submit([=] { std::cout << "second layer" << std::endl; }).get();
    });

    runtime.reset();

    result.get();
```

now it fails here:

```
void consumer_context::destroy() noexcept {
    switch (m_status) {
        case consumer_status::idle: {
            return;
        }

        case consumer_status::await: {
            return storage::destroy(m_storage.caller_handle);
        }

        case consumer_status::wait_for: {
            return storage::destroy(m_storage.wait_for_ctx);
        }

        case consumer_status::when_any: {
            return storage::destroy(m_storage.when_any_ctx);
        }
    }

    assert(false);
}
```

**Comment by Trafo** [comment]
Sorry for giving all the time so less information, because I did not have so much time. I use the latest MSVC compiler, and it seems to be only a problem with MSVC because our macOS application does not have this problem.

The error is very weird because technically the assert(false) should never be hit. In debug mode it shows that m_status has no value anymore, so it is very likely it does not exist anymore at this time point and was cleaned from somewhere else before. I removed the assert out of curiosity, but then bigger problems appear, it fails then inside of a system delete operator. The same happens in the case of using direct release.

I will create a test case for it and open an MR.

**Comment by Trafo** [comment]
The problem is here (include/concurrencpp/results/impl/result_state.h line 49):
```
        static void delete_self(result_state<type>* state) noexcept {
            auto done_handle = state->m_done_handle;
            if (static_cast<bool>(done_handle)) {
                assert(done_handle.done());
                return done_handle.destroy();
            }

            delete state;
        }
```
`done_handle.destroy()` will be called, which is triggering the destructor of `result_state_base`. But the problem is, that after this function call as well the destructor will be called again, but this time on an already destroyed object, which causes the asserts and the release abort. In case of not doing `done_handle.destroy();` at least shutdown works again, because then the destructor will only be called once.

**Comment by Trafo** [comment]
I think something is off. What happening is, that even in case a shutdown is called the inner task will be still executed. During this execution, it realized somehow that it needs to be canceled. Then the `result` type will be destroyed, which has a pointer of the `result_state` in `result_coro_promise`. Because of the `unique_ptr_deleter`, `done_handle.destroy();` will be called in `delete_self` of `result_state`. This destruction method seems to trigger as well `result_coro_promise` dtor. This magic seems to be normal with coroutines, but still don't know why. But now the problem is the same dtor will be called shortly after `destroy` again, why we have a double deletion, which causes `SEGFAULT`.
Could not find out why this happening, because my knowledge regarding coroutines is still too weak.

Sadly I could not find the root cause, I hope you can find it. But at least I found a workaround for me. In case an executor is shut down, it does not spawn a new task anymore. I added checks for the `do_*` functions inside of the executor. Before they do anything, I will check if the shutdown was requested. For that, I refactored a little bit everything.

**Comment by Trafo** [comment]
I was wrong regarding my workaround fix, it only solved it for little edge cases. Made the example more complex and then it throws way more often:

```
#include "concurrencpp/concurrencpp.h"

#include <iostream>

using namespace concurrencpp;

int fibonacci_sync(int i) {
    if (i == 0) {
        return 0;
    }

    if (i == 1) {
        return 1;
    }

    return fibonacci_sync(i - 1) + fibonacci_sync(i - 2);
}

result<int> fibonacci(executor_tag, std::shared_ptr<thread_pool_executor> tpe, const int curr) {
    if (curr <= 10) {
        co_return fibonacci_sync(curr);
    }

    auto fib_1 = fibonacci({}, tpe, curr - 1);
    auto fib_2 = fibonacci({}, tpe, curr - 2);

    co_return co_await fib_1 + co_await fib_2;
}

int main() {
    auto runtime = std::make_shared<concurrencpp::runtime>();
    auto executor = runtime->thread_pool_executor();

    auto fibb_40 = fibonacci({}, executor, 40);

    std::this_thread::sleep_for(std::chrono::milliseconds(20));
    runtime.reset();

    try {
        std::cout << "fibonacci(40) = " << fibb_40.get() << std::endl;
    } catch (...) {
    }

    return 0;
}

```

**Comment by Trafo** [comment]
I understood now, that `result_coro_promise` is part of the coroutine, that's why it gets destroyed after calling `destroy` on the `coroutine_handle<void>` and the rest of the objects as well. The suspicious observation I have is, that when I don't call `destroy` I have obviously a memory leak and the state will not be destructed. In case `destroy` will be called, then the underlying logic tries to call the dtor of `result_coro_promise` twice, or at least something gets triggered to call it the second time. The second time it will cause the problem.

**Comment by Trafo** [comment]
I investigated further and have a question.
After adding more debug prints inside of the code, it becomes more and more of a race if shutdown works or not. I tried to find out, where the difference is between a working shutdown and a failing shutdown. I realized that it always `SEGFAULTS` in case `throw_runtime_shutdown_exception` will be called and thrown. In the other case, it did not happen. In case I disable throwing shutdown and instead early return, shutdown works always. The exception will be `broken_task_exception`.

Do we need the `errors::runtime_shutdown` exception? I assume now that the bad behavior happening during stack unwinding during exception handling. I am sadly not an expert on this topic as well, but it is the only logical explanation I have. From my current understanding removing `errors::runtime_shutdown` and maybe renaming `broken_task` exception to `interrupted_task` exception should be the way to go. `errors::runtime_shutdown` is anyways not well tested in the context of global scope.

**Comment by Trafo** [comment]
I start to understand coroutines more and more. From my perspective, there is no need for the `runtime_shutdown` exception. Especially because in case it crashes, `broken_task` and `runtime_shutdown` will be thrown. Now I can imagine better, how the double destructor call can happen because of the double stack unwind. From my understanding, the shutdown call will empty the queue and in case new things will be added to the queue they will be ignored instead of an exception will be thrown. First, I wanted to introduce a return boolean pattern, but I don't want to introduce overhead.
At the same time, I find `broken_task`, not a good naming in this context and would rename it to `interrupted_task` and stop using `broken_task` and `runtime_shutdown`.
Was testing this approach with the Fibonacci example 10000 times and it does not crash once. Will try to open an MR over the next days.

**Comment by David-Haim** [comment]
OK then, here we go :)

First of all, thanks for reporting the bug and even taking the time trying to debug it. It was a minor bug with an easy fix.

Let's start with things that are not the bug

1. Never create the runtime object with `make_shared`. always create it as a *value type*. you shouldn't destroy the runtime yourself, let the `main` function do its thing.
2. Never pass the runtime object, there is a reason everything inside it is a `shared_ptr`. instead of passing the runtime, pass the executors themselves, they are *already* reference counted, for a reason. 
3. Join your pseudo-main function *before* the main is finished and the runtime is destroyed. 

Combining all of the above, the correct program should be

```cpp

int main() {
    concurrencpp::runtime runtime;
    auto executor = runtime.thread_pool_executor();

    auto fibb_40 = fibonacci({}, executor, 40);

    std::this_thread::sleep_for(std::chrono::milliseconds(20));

    try {
        std::cout << "fibonacci(40) = " << fibb_40.get() << std::endl;
    } catch (...) {
    }

    return 0;
}
```

And also 

```cpp

   concurrencpp::runtime runtime;
    auto thread_executor = runtime.thread_executor();
    auto result = thread_executor->submit([thread_executor] {
        std::cout << "first layer" << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(1));
        thread_executor->submit([=] { std::cout << "second layer" << std::endl; }).get();
    });

    result.get();
```

**Comment by David-Haim** [comment]
Now, for the bug itself.

According to the standard, every coroutine is associated with a `coroutine_handle` which acts as its `this` pointer.
The problem which this object, that besides referencing the coroutine, it does nothing itself. cleaning the coroutine is up to the developer, `coroutine_handle` acts as a C-like pointer and nothing more.

In concurrencpp, the coroutine handle is wrapped by the  [await_via_functor](https://github.com/David-Haim/concurrencpp/blob/master/include/concurrencpp/results/impl/consumer_context.h#L11)  which acts kinda like a `uinque_ptr` for the coroutine handle. 
If the `await_via_functor` is destructed, it resumes the coroutine it's referencing and throws a `broken_task` exception. however, what happens if the coroutine tries to suspend itself, wraps its own `coroutine_handle` with an `await_via_functor` object and while suspending an exception is thrown? the `await_via_functor` destructor will resume the coroutine and will tell it to exit gracefully (by throwing a `broken_task`) then the coroutine is finished.

Now, if we let the exception (that was thrown during the suspension) propagate, it will be thrown to the coroutine we just destroyed! to avoid it, every `await_via_functor` must be wrapped with `try/catch` .
This is what I do in [resume_on](https://github.com/David-Haim/concurrencpp/blob/944a65fc08b33636ac4a1bc3c6fa8e39be1f2fe1/include/concurrencpp/results/resume_on.h#L27) and the [timer_queue](https://github.com/David-Haim/concurrencpp/blob/944a65fc08b33636ac4a1bc3c6fa8e39be1f2fe1/source/timers/timer_queue.cpp#L283).

I did forget wrapping the `await_via_functor` [here (parallel coroutine promise](https://github.com/David-Haim/concurrencpp/blob/944a65fc08b33636ac4a1bc3c6fa8e39be1f2fe1/include/concurrencpp/results/promises.h#L59) and [here (an improvement I made for `executor::enqueue(span))`](https://github.com/David-Haim/concurrencpp/blob/944a65fc08b33636ac4a1bc3c6fa8e39be1f2fe1/include/concurrencpp/executors/executor.h#L37).

It's fixed with the  [latest commit to develop branch](https://github.com/David-Haim/concurrencpp/commit/82855b2934be8c09cf46a1dc0e27ae22b0c2b6d8)

**Comment by Trafo** [comment]
Regarding never using the runtime as shared_ptr/unique_ptr I don't get it what should go wrong regarding that? In my case, I don't have a main function, which can handle the runtime like that. That's why I need to handle it differently.

Thx for referencing the fixes. For me actually, they are not small, because they caused a lot of crashes. :) Besides that, for what `runtime_shutdown` exception is useful? I personally find it more irritating and don't see any value. Because in case I shut down it, my task can finish anyways with `broken_task` or `runtime_shutdown`. Would like it if it is more aligned. At least they have shared the `interrupted_task` exception. Faster it is not as well, because the moment I drop tasks, they destroy themselves and cause direct `broken_tasks`. I actually refactored them all out, but I prefer smaller fixes because better patchable.

Was very interesting, to go through the code. I learned a lot and in parallel did a lot of coroutine tutorials.

**Comment by Trafo** [comment]
It is way better than before and the same issue is not occurring again, but it fails in my test. I wrote this:

```
#include "concurrencpp/concurrencpp.h"

#include <iostream>

using namespace concurrencpp;

int fibonacci_sync(int i) {
    if (i == 0) {
        return 0;
    }

    if (i == 1) {
        return 1;
    }

    return fibonacci_sync(i - 1) + fibonacci_sync(i - 2);
}

result<int> fibonacci(executor_tag, std::shared_ptr<concurrencpp::executor> tpe, const int curr) {
    if (curr <= 10) {
        co_return fibonacci_sync(curr);
    }

    auto fib_1 = fibonacci({}, tpe, curr - 1);
    auto fib_2 = fibonacci({}, tpe, curr - 2);

    co_return co_await fib_1 + co_await fib_2;
}

result<int> simple(executor_tag, std::shared_ptr<concurrencpp::executor> tpe, const int curr) {
    if (curr == 0) {
        co_return 0;
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    auto simple_result = simple({}, tpe, curr - 1);

    co_return co_await simple_result;
}

void t1() {
    auto runtime = std::unique_ptr<concurrencpp::runtime>(new concurrencpp::runtime());
    auto executor = runtime->background_executor();

    auto result = fibonacci({}, executor, 40);

    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    runtime.reset();

    try {
        std::cout << result.get() << std::endl;
    } catch (const std::exception& ex) {
        std::cout << ex.what() << std::endl;
    }
}

void t2() {
    auto runtime = std::unique_ptr<concurrencpp::runtime>(new concurrencpp::runtime());
    auto executor = runtime->inline_executor();

    auto result = simple({}, executor, 15);

    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    runtime.reset();

    try {
        std::cout << result.get() << std::endl;
    } catch (const std::exception& ex) {
        std::cout << ex.what() << std::endl;
    }
}

void t3() {
    auto runtime = std::unique_ptr<concurrencpp::runtime>(new concurrencpp::runtime());
    auto executor = runtime->make_manual_executor();

    auto result = fibonacci({}, executor, 40);

    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    runtime.reset();

    try {
        std::cout << result.get() << std::endl;
    } catch (const std::exception& ex) {
        std::cout << ex.what() << std::endl;
    }
}

void t4() {
    auto runtime = std::unique_ptr<concurrencpp::runtime>(new concurrencpp::runtime());
    auto executor = runtime->make_worker_thread_executor();

    auto result = simple({}, executor, 15);

    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    runtime.reset();

    try {
        std::cout << result.get() << std::endl;
    } catch (const std::exception& ex) {
        std::cout << ex.what() << std::endl;
    }
}

void t5() {
    auto runtime = std::unique_ptr<concurrencpp::runtime>(new concurrencpp::runtime());
    auto executor = runtime->thread_executor();

    auto result = simple({}, executor, 15);

    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    runtime.reset();

    try {
        std::cout << result.get() << std::endl;
    } catch (const std::exception& ex) {
        std::cout << ex.what() << std::endl;
    }
}

void t6() {
    auto runtime = std::unique_ptr<concurrencpp::runtime>(new concurrencpp::runtime());
    auto executor = runtime->thread_pool_executor();

    auto result = fibonacci({}, executor, 40);

    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    runtime.reset();

    try {
        std::cout << result.get() << std::endl;
    } catch (const std::exception& ex) {
        std::cout << ex.what() << std::endl;
    }
}

int main() {
    for (int i = 1; i <= 10000; ++i) {
        std::cout << "attempt: " << i << std::endl;
        t1();
        // t2();
        t3();
        t4();
        t5();
        t6();
    }
    return 0;
}
```

This test runs without any problem when I refactored out `runtime_shutdown` and replaced it with just early returns and some assert adaptions. But with your patch, it aborts during the 40-60 attempt. (without shutdown the full 10000 attempts) In release and debug mode. It seems to have a background executor issue when throwing `runtime_shutdown`.

During `enqueue_foreign` call, the exception will be thrown and the vcruntime140 seems to panic and abort the program. See no issue during creation of the exception.

Besides in the case of  `worker_thread_executor` the `m_private_atomic_abort` is always false. I assume it is a bug. Will it be set inside the shutdown method, `worker_thread_executor` finished way quicker.

**Comment by Trafo** [comment]
Okay, the last bug was easy to find but gives me another reason why the `runtime_shutdown` exception is not a good pattern. The problem is basically that an exception will be thrown in a thread that will nowhere be caught. That can happen in case `shutdown` of `thread_pool_executor` will be called. Then one thread after another will be killed. If during that a task is running `balance_work()` and wants to balance as well, then it could be that a task will be chosen which is already closed. This caused when calling
```
m_parent_pool.worker_at(idle_worker_index).enqueue_foreign(donation_begin_it, donation_end_it);
```
that the exception will be thrown. This exception will nowhere be caught and because it happens in the context of a thread the program will call `terminate()`.
A quick fix is:
```
        try {
            m_parent_pool.worker_at(idle_worker_index).enqueue_foreign(donation_begin_it, donation_end_it);
        } catch (...) {
            return;
        }
```
But it is a dirty one because it assumes this only happens when shutdown is called. Normally the rest of the state needs to be cleaned. But I prefer at least in the case of `thread_pool_executor`, not to use exceptions.

**Comment by Trafo** [comment]
Created an MR. After this fix, my tests don't fail anymore.

Besides that, I want to thank you for this very nice library. It is making C++ so much more powerful, and I like the design. No networking or other stuff is included. At the moment there is no alternative available. Thx for your work.
